reading_comprehension/qanet.py

```diff
@@ -0,0 +1,256 @@
+            and each offset is a token index.
+        loss : torch.FloatTensor, optional
+            A scalar loss to be optimised.
+        best_span_str : List[str]
+            If sufficient metadata was provided for the instances in the batch, we also return the
+            string from the original passage that the model thinks is the best answer to the
+            question.
+        """
+        question_mask = util.get_text_field_mask(question).float()
+        passage_mask = util.get_text_field_mask(passage).float()
+
+        embedded_question = self._dropout(self._text_field_embedder(question))
+        embedded_passage = self._dropout(self._text_field_embedder(passage))
+        embedded_question = self._highway_layer(self._embedding_proj_layer(embedded_question))
+        embedded_passage = self._highway_layer(self._embedding_proj_layer(embedded_passage))
+
+        batch_size = embedded_question.size(0)
+
+        projected_embedded_question = self._encoding_proj_layer(embedded_question)
+        projected_embedded_passage = self._encoding_proj_layer(embedded_passage)
+
+        encoded_question = self._dropout(self._phrase_layer(projected_embedded_question, question_mask))
+        encoded_passage = self._dropout(self._phrase_layer(projected_embedded_passage, passage_mask))
+
+        # Shape: (batch_size, passage_length, question_length)
+        passage_question_similarity = self._matrix_attention(encoded_passage, encoded_question)
+        # Shape: (batch_size, passage_length, question_length)
+        passage_question_attention = masked_softmax(passage_question_similarity, question_mask)
+        # Shape: (batch_size, passage_length, encoding_dim)
+        passage_question_vectors = util.weighted_sum(encoded_question, passage_question_attention)
+
+        # Shape: (batch_size, question_length, passage_length)
+        question_passage_attention = masked_softmax(passage_question_similarity.transpose(1, 2), passage_mask)
+        # Shape: (batch_size, passage_length, passage_length)
+        attention_over_attention = torch.bmm(passage_question_attention, question_passage_attention)
+        # Shape: (batch_size, passage_length, encoding_dim)
+        passage_passage_vectors = util.weighted_sum(encoded_passage, attention_over_attention)
+
+        # Shape: (batch_size, passage_length, encoding_dim * 4)
+        merged_passage_attention_vectors = self._dropout(
+                torch.cat([encoded_passage, passage_question_vectors,
+                           encoded_passage * passage_question_vectors,
+                           encoded_passage * passage_passage_vectors],
+                          dim=-1)
+                )
+
+        modeled_passage_list = [self._modeling_proj_layer(merged_passage_attention_vectors)]
+
+        for _ in range(3):
+            modeled_passage = self._dropout(self._modeling_layer(modeled_passage_list[-1], passage_mask))
```

Comment:
Have you tried removing the masks from these encoder operations?  Does that change anything?

Reply:
Yes, that mask works in the self-attention part.